### ISSUE TITLE
Incomprehensible ValueError thrown by RegionExtractor

### ISSUE THREAD

**Issue by dohmatob** [nilearn/nilearn#907]
I'll find time to expand on the issue desc and provide minimal example to rep.

`RegionExtractor(["image1.nii", "image2.nii"], threshold=3.)`

Throws error:

```
ValueError: threshold given as ratio to the number of voxels must be Real number and should be positive and between 0 and total number of maps i.e. n_maps=2. You provided 3.0
```

**Comment by lesteve** [comment]
Would you mind describing in more details the part of the error message you find incomprehensible. It seems that you have provided threshold=3 and it is telling you you 0. < threshold <= 2.

**Comment by KamalakerDadi** [comment]
I will go into more detailed version of this input type, what happens when user gave a list of 4D imgs. But as far as I see it should be a single 4D Nifti like object not a list of objects. If that's the case, I agree its incomprehensible error message. I will check it anyway.

Do you also have problem with single 4D image ?

**Comment by GaelVaroquaux** [comment]
> But as far as I see it should be a single 4D Nifti like object not a
> list of objects. If that's the case, I agree its incomprehensible error
> message. 

No, I don't think that this is the problem. Nilearn correctly detects
that this is equivalent to a 4D niimg.

The problem is that Elvis didn't understand the bounds of the parameter.

Now that I realize it, that parameter should probably be a fraction, ie
between 0 and 1 independantly of the number of maps input, and it would
be internally multiplied by n_maps.

What do people think?

**Comment by GaelVaroquaux** [comment]
After discussing IRL with @AlexandreAbraham, it seems that it is simply
me who does not understand the error message, and that the
parametrization is correct.

With only 2 maps, it is not possible to allocate voxels to 3 maps on
average.

**Comment by bthirion** [comment]
All this makes sense, but will be suprising to many people who have done neuroimaging with other tools: they are used to equating 'thresholds' and 'statistical thresholds', while the concept of threshold used here is different, something like a 'selection threshold'. Maybe we should give a more explicit naming.

**Comment by GaelVaroquaux** [comment]
> Maybe we should give a more explicit naming.

Yes. I cannot think of one right now, but I agree.

**Comment by bthirion** [comment]
'ratio', 'selection_ratio' ?

**Comment by KamalakerDadi** [comment]
Just to share some information about the usage of threshold and its strategies:

regions extraction based on statistical thresholds can be used based upon selecting ```thresholding_strategy='img_value'``` which will act according to given threshold in statistical value provided that the value should not exceed the ```max```.

regions extraction based on selection threshold can be done upon selecting ```thresholding_strategy='ratio_n_voxels'``` or ```thresholding_strategy='percentile'``` where threshold should be given upon the n_maps provided should not exceed more than n_maps as stated in the error message.

In this particular error case, selecting ```thresholding_strategy='img_value'``` should work if one is trying to threshold upon their statistical values or otherwise leaving to defaults should also work.

We had naming problem before. Since, we have two different strategies, fixing one name from current ```threshold``` to ```ratio``` or ```selection_ratio```  will again be a conflict to statistical thresholding strategy. Let me know if I am wrong.

**Comment by banilo** [comment]
'cutoff'?

**Comment by KamalakerDadi** [comment]
I feel 'cutoff' than 'threshold' is even more difficult to reach to users.

**Comment by KamalakerDadi** [comment]
I am thinking of fixing this issue by rewriting error message as below:

```python 
threshold=3.0 provided to select for number of nonzero voxels across total number of maps (n_maps) is not valid. Please provide valid number between 0. and n_maps=2
```
What do you think ? @dohmatob @lesteve @GaelVaroquaux @AlexandreAbraham @banilo

**Comment by bthirion** [comment]
'number of nonzero voxels across total number of maps (n_maps)' is still unclear to me, because when you say "number of nonzero voxels" people will think "maybe several hundreds (or thousands)".
IMHO the semantics of this parameters would be clearer if it were between 0 and 1, as it would represent a fraction of the population, which is easier to explain. I have to say, it is really hard to figure out what is actually done on the fit().

**Comment by KamalakerDadi** [comment]
> I have to say, it is really hard to figure out what is actually done on the fit(). 

Is it not clear in terms of documentation or implementation ? I am not sure which part is hard to figure ?

**Comment by AlexandreAbraham** [comment]
> IMHO the semantics of this parameters would be clearer if it were between 0 and 1, as it would represent a fraction of the population, which is easier to explain.

It is easier to explain but it defeats the purpose of the method presented here. The main advantage of this thresholding approach is that you can chose a value corresponding to a certain level of sparsity in the brain (1 if you want to avoid overlap, up to 3 if overlaps are OK for you) and then, no matter how many maps you provide, your atlas will always "look" the same, no adjustment needed. When dealing with a lot of atlases, I find it very convenient.

**Comment by bthirion** [comment]
This makes sense, but then we need a serious refactoring of the documentation to clarify this, and ideally an example dedicated to region extraction that illustrates the impact of these parameters.

**Comment by bthirion** [comment]
On 13/01/2016 10:06, KamalakerDadi wrote:
>
>     I have to say, it is really hard to figure out what is actually
>     done on the fit().
>
> Is it not clear in terms of documentation or implementation ? I am not 
> sure which part is hard to figure ?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/nilearn/nilearn/issues/907#issuecomment-171221135>.
>
This means 3 things:
* The code does not run a clear operation on the data, like 
thresholding, mean computation or extraction of connected components or 
even a watershed analysis.
* Handling it with a multi-variate point of view  (n_maps > 1) is also 
non-classical. I'm not aware of any such  procedure apart from that one.
* I am currently trying to run it on other data to build an intuition, 
and do not get it.
Of course I could take time and read the code, but I believe that 
user-level functions should be better explained.
HTH

**Comment by GaelVaroquaux** [comment]
> * Handling it with a multi-variate point of view (n_maps > 1) is also
> non-classical. I'm not aware of any such procedure apart from that one.

Right, but that's exactly what the object is there for.

It should probably be named different, with a name that stresses that it
is multi-maps.

**Comment by KamalakerDadi** [comment]
ping @GaelVaroquaux @AlexandreAbraham @lesteve 

I would like to know your opinion and ideas to bring code to more user friendly and understandable. I am working on to propose some idea. Meanwhile, any ideas or suggestions would be really helpful and great.
